python/adjoint/objective.py

```diff
@@ -231,98 +231,48 @@ def __call__(self):
         super().__init__(sim)
-        self.volume = volume
+        self.volume = sim._fit_volume_to_simulation(volume)
         self.component = component
+        self.yee_grid = yee_grid
         self.decimation_factor = decimation_factor
 
     def register_monitors(self, frequencies):
         self._frequencies = np.asarray(frequencies)
         self._monitor = self.sim.add_dft_fields(
-            [self.component],
-            self._frequencies,
-            where=self.volume,
-            yee_grid=False,
-            decimation_factor=self.decimation_factor,
-        )
+            [self.component], self._frequencies, where=self.volume, yee_grid=self.yee_grid, decimation_factor=self.decimation_factor)
         return self._monitor
 
     def place_adjoint_source(self, dJ):
         time_src = self._create_time_profile()
         sources = []
-        scale = self._adj_src_scale()
 
-        x_dim, y_dim, z_dim = len(self._dg.x), len(self._dg.y), len(self._dg.z)
-
-        if self.num_freq == 1:
-            amp = -dJ[0].copy().reshape(x_dim, y_dim, z_dim) * scale
-            if self.component in [mp.Hx, mp.Hy, mp.Hz]:
-                amp = -amp
-            for zi in range(z_dim):
-                for yi in range(y_dim):
-                    for xi in range(x_dim):
-                        if amp[xi, yi, zi] != 0:
-                            sources += [
-                                mp.Source(
-                                    time_src,
-                                    component=self.component,
-                                    amplitude=amp[xi, yi, zi],
-                                    center=mp.Vector3(self._dg.x[xi],
-                                                      self._dg.y[yi],
-                                                      self._dg.z[zi]),
-                                )
-                            ]
+        mon_size = self.sim.fields.dft_monitor_size(self._monitor.swigobj, self.volume.swigobj, self.component)
+        dJ = dJ.astype(np.complex128)
+        if np.prod(mon_size)*self.num_freq == dJ.size: # The objective function J is a scalar.
+            dJ = dJ.flatten()
+        elif np.prod(mon_size)*self.num_freq**2 == dJ.size: # The objective function J is a vector. Each component corresponds to a frequency.
+            dJ = np.sum(dJ,axis=1).flatten()
         else:
-            '''The adjoint solver requires the objective function
-            to be scalar valued with regard to objective arguments
-            and position, but the function may be vector valued
-            with regard to frequency. In this case, the Jacobian
-            will be of the form [F,F,...] where F is the number of
-            frequencies. Because of linearity, we can sum across the
-            second frequency dimension to calculate a frequency
-            scale factor for each point (rather than a scale vector).
-            '''
```

Comment:
Keep this comment, it's helpful!

Reply:
@smartalecH, could you explain what is going on here?   Why can you sum over frequency if the objective function is vector-valued?